File: space-game/Entities.Tests/EntityManagerTests.vb
Imports NSubstitute
Imports NUnit.Framework

<TestFixture>
Public Class EntityManagerTests

    <TestCase(1, 1)>
    <TestCase(5, 5)>
    <TestCase(20, 20)>
    Public Sub RegisterEntity_WhenCalled_StoresEntity(count As Integer, expected As Integer)

        Dim entityManager = New EntityManager()

        For i = 1 To count
            Dim entity = Substitute.For(Of BaseGameEntity)(entityManager)
        Next

        Assert.AreEqual(expected, entityManager.Count)

    End Sub

    <Test>
    Public Sub RegisterEntity_WhenCalledWithDuplicate_Throws()

        Dim entityManager = New EntityManager()
        Dim expected = GetType(ArgumentException)

        Dim entity = Substitute.For(Of BaseGameEntity)(entityManager)
        Dim ex = Assert.Catch(Of Exception)(Sub() entityManager.RegisterEntity(entity))

        Assert.IsInstanceOf(expected, ex)

    End Sub

    <TestCase(Nothing)>
    Public Sub GetEntityFromId_WhenIdIsNothing_ReturnsNothing(id As Guid)

        Dim entityManager = New EntityManager()

        Dim entity = entityManager.GetEntityFromId(id)

        Assert.AreSame(entity, Nothing)

    End Sub

    <Test>
    Public Sub GetEntityFromId_WhenNoEntityForId_ReturnsNothing()

        Dim entityManager = New EntityManager()
        Dim id As Guid = New Guid()

        Dim entity = entityManager.GetEntityFromId(id)

        Assert.AreSame(entity, Nothing)

    End Sub

    <Test>
    Public Sub GetEntityFromId_WhenIdValid_ReturnsEntity()

        Dim entityManager = New EntityManager()

        Dim entity = Substitute.For(Of BaseGameEntity)(entityManager)
        Dim copyEntity = entityManager.GetEntityFromId(entity.Id)

        Assert.AreSame(entity, copyEntity)

    End Sub


End Class

File: space-game/Entities.Tests/EntityManagerTests.cs
using NSubstitute;
using NUnit.Framework;
using System;

namespace Entities.Tests
{
    [TestFixture]
    public class EntityManagerTests
    {
        [TestCase(1, 1)]
        [TestCase(5, 5)]
        [TestCase(20, 20)]
        public void RegisterEntity_WhenCalled_StoresEntity(int count, int expected)
        {
            EntityManager entityManager = new EntityManager();

            for (int i = 1; i <= count; i++)
            {
                BaseGameEntity entity = Substitute.For<BaseGameEntity>(entityManager);
            }

            Assert.AreEqual(expected, entityManager.Count);
        }

        [Test]
        public void RegisterEntity_WhenCalledWithDuplicate_Throws()
        {
            EntityManager entityManager = new EntityManager();
            Type expected = typeof(ArgumentException);

            BaseGameEntity entity = Substitute.For<BaseGameEntity>(entityManager);
            Exception ex = Assert.Catch<Exception>(() => entityManager.RegisterEntity(entity));

            Assert.IsInstanceOf(expected, ex);
        }

        [TestCase(null)]
        public void GetEntityFromId_WhenIdIsNothing_ReturnsNothing(Guid id)
        {
            EntityManager entityManager = new EntityManager();

            BaseGameEntity entity = entityManager.GetEntityFromId(id);

            Assert.AreSame(entity, null);
        }

        [Test]
        public void GetEntityFromId_WhenNoEntityForId_ReturnsNothing()
        {
            EntityManager entityManager = new EntityManager();
            Guid id = new Guid();

            BaseGameEntity entity = entityManager.GetEntityFromId(id);

            Assert.AreSame(entity, null);
        }

        [Test]
        public void GetEntityFromId_WhenIdValid_ReturnsEntity()
        {
            EntityManager entityManager = new EntityManager();

            BaseGameEntity entity = Substitute.For<BaseGameEntity>(entityManager);
            BaseGameEntity copyEntity = entityManager.GetEntityFromId(entity.Id);

            Assert.AreSame(entity, copyEntity);
        }
    }
}
